<a href="https://colab.research.google.com/github/carlosfmorenog/CM4125/blob/main/CM4125_T7/CM4125_Lab7_2_MeltingPivoting_solved.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CM4125 Topic 7 Lab 2, Melting and Pivoting (Solved)

## Part 1: Melting

Load the datasets `treatment1.csv` and `treatment2.csv` into variables `treatment1` and `treatment2`

On both cases, use `index_col=0`

In [ ]:
import pandas as pd

In [ ]:
url1 = 'https://www.dropbox.com/scl/fi/b0s0v76gv8p3qfmqgs4t7/treatment1.csv?rlkey=h90i0fl8a52h5mv4dqyxrui70&raw=1'
treatment1 = pd.read_csv(url1, index_col=0)
treatment1

treatmenta  treatmentb
John Smith           NaN           2
Jane Doe            16.0          11
Mary Johnson         3.0           1

In [ ]:
url2 = 'https://www.dropbox.com/scl/fi/2zspj08ka40larfjejds8/treatment2.csv?rlkey=j7jy7qedivrzwlhp5483suepb&raw=1'
treatment2 = pd.read_csv(url2, index_col=0)
treatment2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN        16             3
treatmentb         2.0        11             1

**Q. What do you notice is the same between these two datasets?**

Answer: They are the same data, but transposed (i.e. columns and rows are exchanged)

**Q. How many observations are there in these datasets?**

Answer: There are 5 observations if we exclude the null value

Reset the index on `treatment1` so that it now has default indexing, keeping the old index as a column called `patient`

In [ ]:
treatment1 = treatment1.reset_index().rename(columns={'index':'patient'})
treatment1

patient  treatmenta  treatmentb
0    John Smith         NaN           2
1      Jane Doe        16.0          11
2  Mary Johnson         3.0           1

Melt the `treatment1` dataset, and remove the null entry

Make sure you have a table which looks like this (check column headings and data types):

|    | patient      | treatment  | reading |
| -: | :----------- | :--------- | ------: |
|  0 | Jane Doe     | treatmenta |      16 |
|  1 | Mary Johnson | treatmenta |       3 |
|  2 | John Smith   | treatmentb |       2 |
|  3 | Jane Doe     | treatmentb |      11 |
|  4 | Mary Johnson | treatmentb |       1 |

In [ ]:
treatment1 = treatment1.melt(id_vars='patient', var_name='treatment', value_name='reading')
treatment1 = treatment1[treatment1['reading'].notnull()]
treatment1 = treatment1.astype({'reading' : 'int64'})
treatment1 = treatment1.reset_index(drop=True)
treatment1

patient   treatment  reading
0      Jane Doe  treatmenta       16
1  Mary Johnson  treatmenta        3
2    John Smith  treatmentb        2
3      Jane Doe  treatmentb       11
4  Mary Johnson  treatmentb        1

**Q. If we had done the same transformation to `treatment2`, how would the result differ?**

Answer: the data would be the same, although the columns my appear in a different order

## Part 2: Pivoting

The `gdp.csv` dataset shows the gross domestic product (GDP) of counties over time

Load this dataset, using default indexing

In [ ]:
url = 'https://www.dropbox.com/scl/fi/u1nqazwow0qd0denucduq/gdp.csv?rlkey=5ug9po1xv7lmk9tqmoc0t1a7a&raw=1'
gdp = pd.read_csv(url)
gdp

Country Name Country Code     Indicator Name  Indicator Code  \
0           Aruba          ABW  GDP (current US$)  NY.GDP.MKTP.CD   
1     Afghanistan          AFG  GDP (current US$)  NY.GDP.MKTP.CD   
2          Angola          AGO  GDP (current US$)  NY.GDP.MKTP.CD   
3         Albania          ALB  GDP (current US$)  NY.GDP.MKTP.CD   
4         Andorra          AND  GDP (current US$)  NY.GDP.MKTP.CD   
..            ...          ...                ...             ...   
259        Kosovo          XKX  GDP (current US$)  NY.GDP.MKTP.CD   
260   Yemen, Rep.          YEM  GDP (current US$)  NY.GDP.MKTP.CD   
261  South Africa          ZAF  GDP (current US$)  NY.GDP.MKTP.CD   
262        Zambia          ZMB  GDP (current US$)  NY.GDP.MKTP.CD   
263      Zimbabwe          ZWE  GDP (current US$)  NY.GDP.MKTP.CD   

             1960          1961          1962          1963          1964  \
0             NaN           NaN           NaN           NaN           NaN   
1    5.377778e+08  5.488889e+08  5.466667e+08  7.511112e+08  8.000000e+08   
2             NaN           NaN           NaN           NaN           NaN   
3             NaN           NaN           NaN           NaN           NaN   
4             NaN           NaN           NaN           NaN           NaN   
..            ...           ...           ...           ...           ...   
259           NaN           NaN           NaN           NaN           NaN   
260           NaN           NaN           NaN           NaN           NaN   
261  7.575397e+09  7.972997e+09  8.497997e+09  9.423396e+09  1.037400e+10   
262  7.130000e+08  6.962857e+08  6.931429e+08  7.187143e+08  8.394286e+08   
263  1.052990e+09  1.096647e+09  1.117602e+09  1.159512e+09  1.217138e+09   

             1965  ...          2011          2012          2013  \
0             NaN  ...  2.549721e+09  2.534637e+09  2.581564e+09   
1    1.006667e+09  ...  1.780428e+10  2.000162e+10  2.056105e+10   
2             NaN  ...  1.117897e+11  1.280529e+11  1.367099e+11   
3             NaN  ...  1.289087e+10  1.231978e+10  1.277628e+10   
4             NaN  ...  3.442063e+09  3.164615e+09  3.281585e+09   
..            ...  ...           ...           ...           ...   
259           NaN  ...  6.691827e+09  6.499936e+09  7.071960e+09   
260           NaN  ...  3.272642e+10  3.540134e+10  4.041523e+10   
261  1.133440e+10  ...  4.164170e+11  3.963294e+11  3.666449e+11   
262  1.082857e+09  ...  2.346010e+10  2.550337e+10  2.804546e+10   
263  1.311436e+09  ...  1.410192e+10  1.711485e+10  1.909102e+10   

             2014          2015          2016          2017          2018  \
0    2.649721e+09  2.691620e+09  2.646927e+09  2.700559e+09           NaN   
1    2.048487e+10  1.990711e+10  1.936264e+10  2.019176e+10  1.936297e+10   
2    1.457122e+11  1.161936e+11  1.011239e+11  1.221238e+11  1.057510e+11   
3    1.322825e+10  1.138693e+10  1.186135e+10  1.302506e+10  1.505888e+10   
4    3.350736e+09  2.811489e+09  2.877312e+09  3.013387e+09  3.236544e+09   
..            ...           ...           ...           ...           ...   
259  7.386891e+09  6.440612e+09  6.714712e+09  7.227765e+09  7.900269e+09   
260  4.322859e+10  4.262833e+10  3.096824e+10  2.681870e+10  2.691440e+10   
261  3.506376e+11  3.174156e+11  2.963409e+11  3.492681e+11  3.682882e+11   
262  2.715063e+10  2.115439e+10  2.095475e+10  2.586814e+10  2.672007e+10   
263  1.949552e+10  1.996312e+10  2.054868e+10  2.281301e+10  3.100052e+10   

     2019  Unnamed: 64  
0     NaN          NaN  
1     NaN          NaN  
2     NaN          NaN  
3     NaN          NaN  
4     NaN          NaN  
..    ...          ...  
259   NaN          NaN  
260   NaN          NaN  
261   NaN          NaN  
262   NaN          NaN  
263   NaN          NaN  

[264 rows x 65 columns]

Note that this is a very wide data set with 65 columns

The year of the data is stored in each column

Where data is unavailable for the year and country, the entry is missing (`NaN`)

For this exercise, we don't need `'Country Code'`, `'Indicator Name'`, `'Indicator Code'`, or the unnamed column `'Unnamed: 64'`

Remove these from the dataset, so that only the GDP are shown

We only need to keep `'Country name'` and the year-numbered columns

In [ ]:
gdp.drop(columns={'Country Code', 'Indicator Name', 'Indicator Code', 'Unnamed: 64'}, inplace=True)
gdp

Country Name          1960          1961          1962          1963  \
0           Aruba           NaN           NaN           NaN           NaN   
1     Afghanistan  5.377778e+08  5.488889e+08  5.466667e+08  7.511112e+08   
2          Angola           NaN           NaN           NaN           NaN   
3         Albania           NaN           NaN           NaN           NaN   
4         Andorra           NaN           NaN           NaN           NaN   
..            ...           ...           ...           ...           ...   
259        Kosovo           NaN           NaN           NaN           NaN   
260   Yemen, Rep.           NaN           NaN           NaN           NaN   
261  South Africa  7.575397e+09  7.972997e+09  8.497997e+09  9.423396e+09   
262        Zambia  7.130000e+08  6.962857e+08  6.931429e+08  7.187143e+08   
263      Zimbabwe  1.052990e+09  1.096647e+09  1.117602e+09  1.159512e+09   

             1964          1965          1966          1967          1968  \
0             NaN           NaN           NaN           NaN           NaN   
1    8.000000e+08  1.006667e+09  1.400000e+09  1.673333e+09  1.373333e+09   
2             NaN           NaN           NaN           NaN           NaN   
3             NaN           NaN           NaN           NaN           NaN   
4             NaN           NaN           NaN           NaN           NaN   
..            ...           ...           ...           ...           ...   
259           NaN           NaN           NaN           NaN           NaN   
260           NaN           NaN           NaN           NaN           NaN   
261  1.037400e+10  1.133440e+10  1.235500e+10  1.377739e+10  1.489459e+10   
262  8.394286e+08  1.082857e+09  1.264286e+09  1.368000e+09  1.605857e+09   
263  1.217138e+09  1.311436e+09  1.281750e+09  1.397002e+09  1.479600e+09   

     ...          2010          2011          2012          2013  \
0    ...  2.390503e+09  2.549721e+09  2.534637e+09  2.581564e+09   
1    ...  1.585657e+10  1.780428e+10  2.000162e+10  2.056105e+10   
2    ...  8.379950e+10  1.117897e+11  1.280529e+11  1.367099e+11   
3    ...  1.192695e+10  1.289087e+10  1.231978e+10  1.277628e+10   
4    ...  3.355695e+09  3.442063e+09  3.164615e+09  3.281585e+09   
..   ...           ...           ...           ...           ...   
259  ...  5.830464e+09  6.691827e+09  6.499936e+09  7.071960e+09   
260  ...  3.090675e+10  3.272642e+10  3.540134e+10  4.041523e+10   
261  ...  3.753483e+11  4.164170e+11  3.963294e+11  3.666449e+11   
262  ...  2.026556e+10  2.346010e+10  2.550337e+10  2.804546e+10   
263  ...  1.204166e+10  1.410192e+10  1.711485e+10  1.909102e+10   

             2014          2015          2016          2017          2018  \
0    2.649721e+09  2.691620e+09  2.646927e+09  2.700559e+09           NaN   
1    2.048487e+10  1.990711e+10  1.936264e+10  2.019176e+10  1.936297e+10   
2    1.457122e+11  1.161936e+11  1.011239e+11  1.221238e+11  1.057510e+11   
3    1.322825e+10  1.138693e+10  1.186135e+10  1.302506e+10  1.505888e+10   
4    3.350736e+09  2.811489e+09  2.877312e+09  3.013387e+09  3.236544e+09   
..            ...           ...           ...           ...           ...   
259  7.386891e+09  6.440612e+09  6.714712e+09  7.227765e+09  7.900269e+09   
260  4.322859e+10  4.262833e+10  3.096824e+10  2.681870e+10  2.691440e+10   
261  3.506376e+11  3.174156e+11  2.963409e+11  3.492681e+11  3.682882e+11   
262  2.715063e+10  2.115439e+10  2.095475e+10  2.586814e+10  2.672007e+10   
263  1.949552e+10  1.996312e+10  2.054868e+10  2.281301e+10  3.100052e+10   

     2019  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN  
..    ...  
259   NaN  
260   NaN  
261   NaN  
262   NaN  
263   NaN  

[264 rows x 61 columns]

Convert the data set to long format with columns `'Country Name'`, `'Year'`, and `'GDP'` as `id_vars`, `var_name` and `value_name` respectively.

Here is an example of the first three rows of the data

|    | Country Name | Year |          GDP |
| -: | :----------- | ---: | -----------: |
| 0  | Aruba        | 1960 |          NaN |
| 1  | Afghanistan  | 1960 | 5.377778e+08 |
| 2  | Angola       | 1960 |          NaN |

In [ ]:
gdp = gdp.melt(id_vars = ['Country Name'],
               var_name = 'Year',
               value_name = 'GDP')
gdp

Country Name  Year           GDP
0             Aruba  1960           NaN
1       Afghanistan  1960  5.377778e+08
2            Angola  1960           NaN
3           Albania  1960           NaN
4           Andorra  1960           NaN
...             ...   ...           ...
15835        Kosovo  2019           NaN
15836   Yemen, Rep.  2019           NaN
15837  South Africa  2019           NaN
15838        Zambia  2019           NaN
15839      Zimbabwe  2019           NaN

[15840 rows x 3 columns]

Remove the null entries in the dataset

In [ ]:
gdp = gdp[gdp['GDP'].notnull()]
gdp

Country Name  Year           GDP
1       Afghanistan  1960  5.377778e+08
11        Australia  1960  1.857767e+10
12          Austria  1960  6.592694e+09
14          Burundi  1960  1.960000e+08
15          Belgium  1960  1.165872e+10
...             ...   ...           ...
15571        Kosovo  2018  7.900269e+09
15572   Yemen, Rep.  2018  2.691440e+10
15573  South Africa  2018  3.682882e+11
15574        Zambia  2018  2.672007e+10
15575      Zimbabwe  2018  3.100052e+10

[12089 rows x 3 columns]

**Q. How many observations do we have in the data set?**

Answer: 12089

Create a new dataframe in a variable `uk`, which contains the GDP for only the United Kingdom, with `'Year'` as the index

Since the numbers are quite large, divide by `1_000_000_000_000` (one trillion), and round to 3 decimal places

Display the most recent $5$ years available in the data, as follows:

|     Year | GDP (Trillion USD) |
| -------: | -----------------: |
| **2014** |              3.035 |
| **2015** |              2.896 |
| **2016** |              2.659 |
| **2017** |              2.638 |
| **2018** |              2.825 |

In [ ]:
uk = gdp[gdp['Country Name'] == 'United Kingdom']
uk = uk.drop(columns=['Country Name'])
uk = uk.set_index('Year')
uk['GDP (Trillion USD)'] = (uk['GDP'] / 1_000_000_000_000).round(3)
uk = uk.drop(columns={'GDP'})
uk.tail(5)

GDP (Trillion USD)
Year                    
2014               3.035
2015               2.896
2016               2.659
2017               2.638
2018               2.825

Let's go back to the original dataframe `gdp`

You should have this data set now in the "long" format

Change the table back to the wide format using `pivot`, this time instead of having a column for each `Year`, have a column for each `Country`, the `Year` will be the rows

In [ ]:
gdp.pivot(index='Year', columns='Country Name', values='GDP')

Country Name   Afghanistan       Albania       Algeria  American Samoa  \
Year                                                                     
1960          5.377778e+08           NaN  2.723649e+09             NaN   
1961          5.488889e+08           NaN  2.434777e+09             NaN   
1962          5.466667e+08           NaN  2.001469e+09             NaN   
1963          7.511112e+08           NaN  2.703015e+09             NaN   
1964          8.000000e+08           NaN  2.909352e+09             NaN   
1965          1.006667e+09           NaN  3.136259e+09             NaN   
1966          1.400000e+09           NaN  3.039835e+09             NaN   
1967          1.673333e+09           NaN  3.370843e+09             NaN   
1968          1.373333e+09           NaN  3.852116e+09             NaN   
1969          1.408889e+09           NaN  4.257219e+09             NaN   
1970          1.748887e+09           NaN  4.863487e+09             NaN   
1971          1.831109e+09           NaN  5.077222e+09             NaN   
1972          1.595555e+09           NaN  6.761786e+09             NaN   
1973          1.733333e+09           NaN  8.715106e+09             NaN   
1974          2.155555e+09           NaN  1.320971e+10             NaN   
1975          2.366667e+09           NaN  1.555793e+10             NaN   
1976          2.555556e+09           NaN  1.772835e+10             NaN   
1977          2.953333e+09           NaN  2.097190e+10             NaN   
1978          3.300000e+09           NaN  2.636449e+10             NaN   
1979          3.697940e+09           NaN  3.324342e+10             NaN   
1980          3.641723e+09           NaN  4.234528e+10             NaN   
1981          3.478788e+09           NaN  4.434867e+10             NaN   
1982                   NaN           NaN  4.520709e+10             NaN   
1983                   NaN           NaN  4.880137e+10             NaN   
1984                   NaN  1.857338e+09  5.369828e+10             NaN   
1985                   NaN  1.897050e+09  5.793787e+10             NaN   
1986                   NaN  2.097326e+09  6.369630e+10             NaN   
1987                   NaN  2.080796e+09  6.674227e+10             NaN   
1988                   NaN  2.051236e+09  5.908907e+10             NaN   
1989                   NaN  2.253090e+09  5.563149e+10             NaN   
1990                   NaN  2.028554e+09  6.204510e+10             NaN   
1991                   NaN  1.099559e+09  4.571537e+10             NaN   
1992                   NaN  6.521750e+08  4.800330e+10             NaN   
1993                   NaN  1.185315e+09  4.994646e+10             NaN   
1994                   NaN  1.880952e+09  4.254257e+10             NaN   
1995                   NaN  2.392765e+09  4.176405e+10             NaN   
1996                   NaN  3.199643e+09  4.694150e+10             NaN   
1997                   NaN  2.258516e+09  4.817786e+10             NaN   
1998                   NaN  2.545967e+09  4.818775e+10             NaN   
1999                   NaN  3.212119e+09  4.864057e+10             NaN   
2000                   NaN  3.480355e+09  5.479025e+10             NaN   
2001                   NaN  3.922099e+09  5.474471e+10             NaN   
2002          4.055177e+09  4.348070e+09  5.676029e+10     514000000.0   
2003          4.515563e+09  5.611492e+09  6.786383e+10     527000000.0   
2004          5.226775e+09  7.184681e+09  8.532500e+10     512000000.0   
2005          6.209140e+09  8.052076e+09  1.031982e+11     503000000.0   
2006          6.971287e+09  8.896074e+09  1.170273e+11     496000000.0   
2007          9.747886e+09  1.067732e+10  1.349771e+11     520000000.0   
2008          1.010922e+10  1.288135e+10  1.710007e+11     563000000.0   
2009          1.243909e+10  1.204421e+10  1.372110e+11     678000000.0   
2010          1.585657e+10  1.192695e+10  1.612073e+11     576000000.0   
2011          1.780428e+10  1.289087e+10  2.000191e+11     574000000.0   
2012

**Q. What is the shape of this data frame?**

Answer: 59 rows × 258 columns

Returning again to the long version of the dataset, use `pivot_table` to find the maximum (`max`) GDP for each country (the highest the GDP has ever been in the recorded data)

Store this in a variable called `highest_ever`

Should look something like this


| Country Name |          GDP |
| :----------- | -----------: |
| Afghanistan  | 2.056105e+10 |
| Albania      | 1.505888e+10 |
| Algeria      | 2.138100e+11 |

*Hint: Do not specify a `columns=` parameter, since we want only a single column.*

In [ ]:
highest_ever = gdp.pivot_table(index='Country Name', values='GDP', aggfunc='max')
highest_ever

GDP
Country Name                    
Afghanistan         2.056105e+10
Albania             1.505888e+10
Algeria             2.138100e+11
American Samoa      6.780000e+08
Andorra             4.016972e+09
...                          ...
West Bank and Gaza  1.461590e+10
World               8.580439e+13
Yemen, Rep.         4.322859e+10
Zambia              2.804546e+10
Zimbabwe            3.100052e+10

[258 rows x 1 columns]

**Q. Which country has had the highest ever GDP?**

*Hint: The dataset contains entries which are not countries, e.g. `World` or `High income`. If you display the first 20 rows after sorting, you should be able to find the answer.*

In [ ]:
highest_ever.sort_values('GDP', ascending=False).head(20)

GDP
Country Name                                             
World                                        8.580439e+13
High income                                  5.411879e+13
OECD members                                 5.261014e+13
Post-demographic dividend                    5.002770e+13
IDA & IBRD total                             3.267245e+13
Low & middle income                          3.171682e+13
Middle income                                3.114302e+13
IBRD only                                    3.046677e+13
East Asia & Pacific                          2.591829e+13
Upper middle income                          2.443513e+13
Europe & Central Asia                        2.370681e+13
North America                                2.221325e+13
Late-demographic dividend                    2.217029e+13
United States                                2.049410e+13
European Union                               1.916362e+13
East Asia & Pacific (excluding high income)  1.628000e+13
East Asia & Pacific (IDA & IBRD countries)   1.624877e+13
Euro area                                    1.412488e+13
China                                        1.360815e+13
Early-demographic dividend                   1.163883e+13

Answer: United States

**Extra Question: Is there any way in which we can better filter this dataset to look for countries instead of regions?**

Answer: One idea is to define a dictionary of countries and filter out entries from the `highest_ever` dataframe. Another option is that if we find a list of questions on a csv online, we can import it and then do a join between both tables so that the left is this list, and the right is `highest_ever`. If we do a left join, then only countries will remain!

Or **harness** the power of AI!

In [ ]:
## Prompts Used:
# 1. create code to find the country that has the highest GDP value in the highest_ever pandas dataframe
# 2. some of the entries in the index are not countries, how to find the country?

## Result
# List of known aggregate entries to exclude
aggregate_entries = ['World', 'High income', 'OECD members', 'Post-demographic dividend', 'IDA & IBRD total',
                     'Low & middle income', 'Middle income', 'IBRD only', 'East Asia & Pacific', 'Upper middle income',
                     'Europe & Central Asia', 'North America', 'Late-demographic dividend', 'European Union',
                     'East Asia & Pacific (excluding high income)', 'East Asia & Pacific (IDA & IBRD countries)',
                     'Euro area', 'Early-demographic dividend']

# Filter out aggregate entries
filtered_highest_ever = highest_ever[~highest_ever.index.isin(aggregate_entries)]

# Find the country with the highest GDP from the filtered data
country_with_highest_gdp_filtered = filtered_highest_ever.sort_values('GDP', ascending=False).index[0]

print(f"The country with the highest ever GDP (excluding aggregate entries) is: {country_with_highest_gdp_filtered}")

The country with the highest ever GDP (excluding aggregate entries) is: United States
